pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/BaseSegmentAssignment.java

```diff
@@ -116,33 +107,11 @@ protected List<String> assignSegment(String segmentName, Map<String, Map<String,
     int numReplicaGroups = instancePartitions.getNumReplicaGroups();
     int numPartitions = instancePartitions.getNumPartitions();
 
-    if (numReplicaGroups == 1 && numPartitions == 1) {
-      // Non-replica-group based assignment
-
-      return SegmentAssignmentUtils.assignSegmentWithoutReplicaGroup(currentAssignment, instancePartitions,
-          _replication);
-    } else {
-      // Replica-group based assignment
-
-      checkReplication(instancePartitions);
-
-      int partitionId;
-      if (_partitionColumn == null || numPartitions == 1) {
-        partitionId = 0;
-      } else {
-        // Uniformly spray the segment partitions over the instance partitions
-        partitionId = getSegmentPartitionId(segmentName) % numPartitions;
-      }
-
-      return SegmentAssignmentUtils.assignSegmentWithReplicaGroup(currentAssignment, instancePartitions, partitionId);
-    }
+    // determine segment assignment strategy
+    setSegmentAssignmentStrategy(numReplicaGroups, numPartitions);
```

Comment:
`assignSegment()` will be invoked multiple times after we generate the `SegmentAssignment`object. 

Instead of creating `AssignmentStrategy` everytime we invoke this function, we should re-use the assignment strategy object. Can you think of the way to achieve this?

Reply:
So I moved determining segment assignment strategy to PinotHelixResourceManager where we also determine SegmentAssignment object to avoid recalculations in assign segment, For rebalance as well, moved determining segment assignment strategy to TableRebalancer and pass this assignmentStrategyObject to segment Assignment.

WDYT?